From 7946d213283a138c16958e84a156c6b2ff939204 Mon Sep 17 00:00:00 2001
From: Nicolas Boichat <drinkcat@google.com>
Date: Mon, 16 May 2016 14:23:18 +0800
Subject: [PATCH] pvr_dri: Route logging messages to Android logcat

On Android, we lose the information if it is printed to stderr.

BUG=chrome-os-partner:53338
TEST=N/A

Signed-off-by: Nicolas Boichat <drinkcat@google.com>
---
 src/mesa/drivers/dri/pvr/Makefile.am |  4 ++++
 src/mesa/drivers/dri/pvr/pvrutil.c   | 21 ++++++++++++++++++---
 2 files changed, 22 insertions(+), 3 deletions(-)

diff --git a/src/mesa/drivers/dri/pvr/Makefile.am b/src/mesa/drivers/dri/pvr/Makefile.am
index b6f76ea..1fbd8dd 100644
--- a/src/mesa/drivers/dri/pvr/Makefile.am
+++ b/src/mesa/drivers/dri/pvr/Makefile.am
@@ -17,6 +17,10 @@ AM_CFLAGS = \
 AM_CFLAGS += -DPVR_OGLES1_LIB=\"libGLESv1_CM_POWERVR_ROGUE.so\"
 AM_CFLAGS += -DPVR_OGLES2_LIB=\"libGLESv2_POWERVR_ROGUE.so\"
 
+if HAVE_EGL_PLATFORM_ANDROID
+AM_CFLAGS += -DHAVE_ANDROID_PLATFORM
+endif
+
 AM_CXXFLAGS = $(AM_CFLAGS)
 
 noinst_LTLIBRARIES = libpvr_dri.la
diff --git a/src/mesa/drivers/dri/pvr/pvrutil.c b/src/mesa/drivers/dri/pvr/pvrutil.c
index e020214..7db43b8 100644
--- a/src/mesa/drivers/dri/pvr/pvrutil.c
+++ b/src/mesa/drivers/dri/pvr/pvrutil.c
@@ -37,6 +37,16 @@ THE SOFTWARE.
 
 #include "pvrdri.h"
 
+#ifdef HAVE_ANDROID_PLATFORM
+#define LOG_TAG "PVR-MESA"
+#include <cutils/log.h>
+#define err_vprintf(f, args) LOG_PRI_VA(ANDROID_LOG_ERROR, LOG_TAG, f, args)
+#define dbg_vprintf(f, args) LOG_PRI_VA(ANDROID_LOG_DEBUG, LOG_TAG, f, args)
+#else
+#define err_vprintf(f, args) vfprintf(stderr, f, args)
+#define dbg_vprintf(f, args) vfprintf(stderr, f, args)
+#endif /* HAVE_ANDROID_PLATFORM */
+
 static const PVRDRIImageFormat g_asFormats[] =
 {
 	{
@@ -220,14 +230,13 @@ static const PVRDRIImageFormat g_asFormats[] =
 #endif
 };
 
-
 /* Standard error message */
 void __attribute__((format(printf, 1, 2))) errorMessage(const char *f, ...)
 {
 	va_list args;
 
 	va_start(args, f);
-	vfprintf(stderr, f, args);
+	err_vprintf(f, args);
 	va_end(args);
 }
 
@@ -236,15 +245,21 @@ void __attribute__((format(printf, 1, 2))) __driUtilMessage(const char *f, ...)
 	va_list args;
 	char *ev;
 
+#ifdef HAVE_ANDROID_PLATFORM
+	va_start(args, f);
+	dbg_vprintf(f, args);
+	va_end(args);
+#else
 	ev = getenv("LIBGL_DEBUG");
 	if (ev != NULL && (strcmp(ev, "verbose") == 0))
 	{
 		fputs("LibGL: ", stderr);
 		va_start(args, f);
-		vfprintf(stderr, f, args);
+		dbg_vprintf(stderr, f, args);
 		va_end(args);
 		fputc('\n', stderr);
 	}
+#endif
 }
 
 void *PVRDRILoadAPILib(const char *name)
-- 
2.8.0.rc3.226.g39d4020

